docs(hello-world): add-system and add-table (#1763)
latticexyz/mud: next-docs/pages/guides/hello-world/add-table.mdx

   // (See IncrementSystem.sol)
   (window as any).increment = async () => {
     console.log("new counter value:", await increment());
   };

   (window as any).readHistory = async (counterValue) => {
     const History = components.History;
     const entity = encodeEntity(History.metadata.keySchema, { counterValue });
     const { blockNumber, time } = getComponentValueStrict(History, entity);
     document.getElementById("blockNumber")!.innerHTML = blockNumber;
     document.getElementById("timeStamp")!.innerHTML = new Date(parseInt(Number(time) * 1000));
   };

   if (import.meta.env.DEV) {
     const { mount: mountDevTools } = await import("@latticexyz/dev-tools");
     mountDevTools({
       config: mudConfig,
       publicClient: network.publicClient,
       walletClient: network.walletClient,
       latestBlock$: network.latestBlock$,
       blockStorageOperations$: network.blockStorageOperations$,
       worldAddress: network.worldContract.address,
       worldAbi: network.worldContract.abi,
       write$: network.write$,
       recsWorld: network.world,
     });
   }
   ```

   </CollapseCode>

<details>
<summary>Explanation</summary>

```typescript

> let options: String = "";

Reviewer: ```suggestion
let options: string = "";
```

nit, let's use the primitive type over the class (see https://www.geeksforgeeks.org/what-is-the-difference-between-string-and-string-in-typescript)

Reviewer: in fact we can remove `: string` since we're initializing it in the same line, so typescript can infer the type

```

Create `options` as an empty string.
This is the way you define a variable in TypeScript: `let <variable name>: <type>`.
Here we initialize it to the empty string.

```typescript
for (let i = 1; i <= nextValue?.value; i++) {
  options += `<option value="${i}">${i}</option>`;
}
```

Create the list of options.

```typescript
document.getElementById("historyValue")!.innerHTML = options;
```

Set the internal HTML of the `historyValue` HTML tag to `options`.
Notice the exclamation mark (`!`).
`document.getElementById` may return either a tag that can be changed, or an empty value (if the parameter is not an id of any of the HTML tags).
We know that `historyValue` exists in the HTML, but the TypeScript compiler does not.
This exclamation point tells the compiler that it's OK, there will be a real value there.
[See here for additional information](https://blog.logrocket.com/understanding-exclamation-mark-typescript/).

```typescript
(window as any).readHistory = async counterValue => {
  const History = components.History
  const entity = encodeEntity(History.metadata.keySchema, { counterValue });
```

> `encodeEntity` creates a key in the format that MUD uses, which is based on [ABI argument encoding](https://docs.soliditylang.org/en/latest/abi-spec.html#argument-encoding).

Reviewer: ```suggestion
`encodeEntity` creates an entity key with the [encoding that MUD uses](/store/encoding).
```

we can link to the MUD encoding docs page here

```typescript
const { blockNumber, time } = getComponentValueStrict(History, entity);
```

Read the actual data.

```typescript
document.getElementById("blockNumber")!.innerHTML = blockNumber;
```

Update the value in the HTML table.

```typescript
  document.getElementById("timeStamp")!.innerHTML =
    new Date(parseInt(Number(time) * 1000))
}
```

Solidity uses [Unix time](https://en.wikipedia.org/wiki/Unix_time).
JavaScript uses a similar system, but it measures times in milliseconds.

> So to get a readable date, we take the time (which is a [`BigInt`](https://www.w3schools.com/js/js_bigint.asp)), multiply it by a thousand, and then convert it to a [`Date`](https://www.w3schools.com/jsref/jsref_obj_date.asp) object.

Reviewer: ```suggestion
So to get a readable date, we take the time (which is a [`BigInt`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt)), multiply it by a thousand, and then convert it to a [`Date`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date) object.
```

</details>

2. Edit `packages/clients/index.html`.

   <CollapseCode>

   ```html filename="index.html" copy showLineNumbers {12-26}
   <!DOCTYPE html>
   <html lang="en">
     <head>
       <meta charset="UTF-8" />
       <meta name="viewport" content="width=device-width, initial-scale=1.0" />
       <title>a minimal MUD client</title>
     </head>
     <body>
       <script type="module" src="/src/index.ts"></script>
       <div>Counter: <span id="counter">0</span></div>
       <button onclick="window.increment()">Increment</button>
       <hr />
       <h2>
         History for value:
         <select id="historyValue" onInput="window.readHistory(value)"></select>
       </h2>
       <table border>
         <tr>
           <th>Block number</th>
           <th>Time</th>
         </tr>
         <tr>
           <td id="blockNumber"></td>
           <td id="timeStamp"></td>
         </tr>
       </table>
     </body>
   </html>
   ```

   </CollapseCode>

<details>
<summary>Explanation</summary>

```html
<hr />
<h2>
  History for value:
  <select id="historyValue" onInput="window.readHistory(value)"></select>
</h2>
```

> This is the [input field](https://www.w3schools.com/tags/tag_select.asp) that lets the user select which counter value they'd like to get information about.

Reviewer: ```suggestion
This is the [input field](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select) that lets the user select which counter value they'd like to get information about.
```

The `id` attribute is used by `packages/client/src/index.ts` to set the options.
The `onInput` attribute is the JavaScript code to execute when the value changes.

```html
<table border></table>
```

> A standard [HTML table](https://www.w3schools.com/html/html_tables.asp).

Reviewer: ```suggestion
A standard [HTML table](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/table).
```
